src/librustc/middle/ty/structural_impls.rs

```diff
@@ -751,6 +658,18 @@ impl<'tcx> TypeFoldable<'tcx> for ty::Predicate<'tcx> {
                 ty::Predicate::ObjectSafe(trait_def_id),
         }
     }
```

Comment:
@jseyfried 

> Did you want fold_subitems_with to be unimplemented! to reflect that users should only be calling fold_subitems_with when it is different from fold_with?

I think my reasoning was that outside users should *always* call `fold_with`, and the only reason to have `fold_subitems_with` was to use from the visitor method itself. I'm trying to think if there would be a correct scenario for calling `fold_subitems_with` on one of these cases where we haven't allowed for overriding in the visitor.

I think I agree with this comment though:

> On second though, I think it would nicest to have fold_with default to fold_subitems_with so that structural implementations always define fold_subitem_with

...in the MIR visitor, I actually used the name `super_fold_with`, which I personally find more intuitive than `fold_subitems_with`. It also makes it clear that this is not something you normally want to call. Otherwise my setup was just as you described:

https://github.com/rust-lang/rust/blob/master/src/librustc/mir/visit.rs#L21-L23

Would you mind refactoring to use the `super` naming scheme and to reverse the defaults?